From bc4bdd8b5aafb7165b182cb20554aaebbae768b8 Mon Sep 17 00:00:00 2001
From: Sun Yuechi <sunyuechi@iscas.ac.cn>
Date: Sat, 12 Sep 2026 10:21:10 -0700
Subject: [PATCH] cmake: link rgw, crimson and ceph-osd in the heavy job pool

These targets link the rgw or crimson static stack; they and ceph-osd
peak higher in ld than unittest_librbd, which is already in
heavy_link_job_pool (x86-64 Debug: radosgw 2.2 GB, ceph-osd 1.7 GB,
unittest_librbd 1.4 GB). Linking many of them at once ran out of memory
in an ASan build on a 64-core, 128 GB riscv64 host.

Signed-off-by: Sun Yuechi <sunyuechi@iscas.ac.cn>
---
 src/CMakeLists.txt                     | 2 ++
 src/crimson/osd/CMakeLists.txt         | 3 ++-
 src/crimson/tools/CMakeLists.txt       | 2 ++
 src/rgw/CMakeLists.txt                 | 2 ++
 src/test/CMakeLists.txt                | 5 +++++
 src/test/crimson/CMakeLists.txt        | 2 ++
 src/test/rgw/CMakeLists.txt            | 2 ++
 src/tools/ceph-dencoder/CMakeLists.txt | 2 ++
 8 files changed, 19 insertions(+), 1 deletion(-)

diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index fb07314fd3d..0f405a90fbb 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -839,6 +839,8 @@ set(ceph_osd_srcs
   ceph_osd.cc)
 
 add_executable(ceph-osd ${ceph_osd_srcs})
+set_target_properties(ceph-osd PROPERTIES
+  JOB_POOL_LINK heavy_link_job_pool)
 add_dependencies(ceph-osd erasure_code_plugins extblkdev_plugins)
 target_link_libraries(ceph-osd osd os global-static common
   ${ALLOC_LIBS}
diff --git a/src/crimson/osd/CMakeLists.txt b/src/crimson/osd/CMakeLists.txt
index a15adf049cc..cdac5700337 100644
--- a/src/crimson/osd/CMakeLists.txt
+++ b/src/crimson/osd/CMakeLists.txt
@@ -101,7 +101,8 @@ else()
     dmclock::dmclock)
 endif()
 set_target_properties(crimson-osd PROPERTIES
-  POSITION_INDEPENDENT_CODE ${EXE_LINKER_USE_PIE})
+  POSITION_INDEPENDENT_CODE ${EXE_LINKER_USE_PIE}
+  JOB_POOL_LINK heavy_link_job_pool)
 install(TARGETS crimson-osd DESTINATION bin)
 if(WITH_TESTS)
   add_dependencies(tests crimson-osd)
diff --git a/src/crimson/tools/CMakeLists.txt b/src/crimson/tools/CMakeLists.txt
index 6898efa04a0..e05eccc6525 100644
--- a/src/crimson/tools/CMakeLists.txt
+++ b/src/crimson/tools/CMakeLists.txt
@@ -1,3 +1,5 @@
+set(CMAKE_JOB_POOL_LINK heavy_link_job_pool)
+
 add_executable(crimson-store-nbd
 	store_nbd/store-nbd.cc
 	store_nbd/tm_driver.cc
diff --git a/src/rgw/CMakeLists.txt b/src/rgw/CMakeLists.txt
index f22470840e5..b71dbd3b621 100644
--- a/src/rgw/CMakeLists.txt
+++ b/src/rgw/CMakeLists.txt
@@ -1,3 +1,5 @@
+set(CMAKE_JOB_POOL_LINK heavy_link_job_pool)
+
 find_program(GPERF gperf
   REQUIRED)
 
diff --git a/src/test/CMakeLists.txt b/src/test/CMakeLists.txt
index a67b0e0209f..4e9480fc12a 100644
--- a/src/test/CMakeLists.txt
+++ b/src/test/CMakeLists.txt
@@ -137,6 +137,8 @@ target_link_libraries(test_build_libcommon ceph-common pthread ${CRYPTO_LIBS} ${
 if(WITH_RADOSGW)
   add_executable(test_build_librgw buildtest_skeleton.cc)
   target_link_libraries(test_build_librgw rgw_a pthread ${CRYPTO_LIBS} ${EXTRALIBS})
+  set_target_properties(test_build_librgw PROPERTIES
+    JOB_POOL_LINK heavy_link_job_pool)
 endif(WITH_RADOSGW)
 
 if(WITH_LIBCEPHFS)
@@ -299,6 +301,9 @@ if(${WITH_RADOSGW})
     ${EXTRALIBS}
     ${CRYPTO_LIBS}
     )
+  set_target_properties(test_cors ceph_test_cls_rgw_admin_account_quota_set
+    ceph_test_cls_rgw_meta ceph_test_cls_rgw_log PROPERTIES
+    JOB_POOL_LINK heavy_link_job_pool)
 
 # ceph_test_librgw_file (nfs-like RGW interface)
 add_executable(ceph_test_librgw_file
diff --git a/src/test/crimson/CMakeLists.txt b/src/test/crimson/CMakeLists.txt
index 71e0b876f67..4c58da51eee 100644
--- a/src/test/crimson/CMakeLists.txt
+++ b/src/test/crimson/CMakeLists.txt
@@ -1,3 +1,5 @@
+set(CMAKE_JOB_POOL_LINK heavy_link_job_pool)
+
 # the crimson's backfill doesn't need nor use seastar
 include_directories(${CMAKE_SOURCE_DIR}/src/dmclock/support/src)
 add_executable(unittest-crimson-backfill
diff --git a/src/test/rgw/CMakeLists.txt b/src/test/rgw/CMakeLists.txt
index 8e70a21833c..fcb98b9262e 100644
--- a/src/test/rgw/CMakeLists.txt
+++ b/src/test/rgw/CMakeLists.txt
@@ -1,3 +1,5 @@
+set(CMAKE_JOB_POOL_LINK heavy_link_job_pool)
+
 #
 # Helper for adding RGW tests built with Catch2:
 # Note: to define a test that does not use Catch2::main(), use:
diff --git a/src/tools/ceph-dencoder/CMakeLists.txt b/src/tools/ceph-dencoder/CMakeLists.txt
index f0a3f81ea1b..00888302b0a 100644
--- a/src/tools/ceph-dencoder/CMakeLists.txt
+++ b/src/tools/ceph-dencoder/CMakeLists.txt
@@ -62,6 +62,8 @@ if(WITH_RADOSGW)
     rgw_a
     cls_rgw_client
     cls_journal_client)
+  set_target_properties(denc-mod-rgw PROPERTIES
+    JOB_POOL_LINK heavy_link_job_pool)
   if(WITH_RADOSGW_AMQP_ENDPOINT)
     target_link_libraries(denc-mod-rgw
       rabbitmq ssl)
-- 
2.55.0

